File: shared/config-ts/src/storage.ts
/**
 * Config file storage with atomic writes and file locking.
 */

import { readFileSync, writeFileSync, renameSync, existsSync } from 'node:fs';
import lockfile from 'proper-lockfile';
import { ConfigSchema } from './schemas.js';
import { ConfigCorruptError } from './errors.js';
import { getConfigFile, getLockFile } from './paths.js';
import { Config, createDefaultConfig } from './types.js';

// Lock timeout in milliseconds
const LOCK_TIMEOUT = 5000;

/**
 * Read the configuration file.
 *
 * If the file doesn't exist, returns default config.
 * If the file is corrupted, logs a warning and returns default config.
 */
export function readConfig(): Config {
  const configFile = getConfigFile(true);

  if (!existsSync(configFile)) {
    return createDefaultConfig();
  }

  try {
    const data = JSON.parse(readFileSync(configFile, 'utf-8'));
    const result = ConfigSchema.safeParse(data);

    if (!result.success) {
      console.warn(`Config validation failed: ${result.error.message}`);
      return createDefaultConfig();
    }

    return result.data;
  } catch (e) {
    if (e instanceof SyntaxError) {
      console.warn(`Config file corrupted (invalid JSON): ${e.message}`);
      return createDefaultConfig();
    }
    throw e;
  }
}

/**
 * Read config synchronously (alias for readConfig).
 */
export function readConfigSync(): Config {
  return readConfig();
}

/**
 * Write the configuration file atomically with locking.
 */
export async function writeConfig(config: Config): Promise<void> {
  const configFile = getConfigFile(true);

  // Use proper-lockfile for cross-platform locking
  const release = await lockfile.lock(configFile, {
    retries: { retries: 3, minTimeout: 100, maxTimeout: 1000 },
    stale: LOCK_TIMEOUT,
    realpath: false,
  }).catch(() => {
    // If file doesn't exist yet, create it first
    if (!existsSync(configFile)) {
      writeFileSync(configFile, JSON.stringify(createDefaultConfig(), null, 2));
      return lockfile.lock(configFile, {
        retries: { retries: 3, minTimeout: 100, maxTimeout: 1000 },
        stale: LOCK_TIMEOUT,
        realpath: false,
      });
    }
    throw new Error(`Could not acquire lock on ${configFile}`);
  });

  try {
    // Write to temp file first
    const tempFile = configFile.replace('.json', '.tmp');
    writeFileSync(tempFile, JSON.stringify(config, null, 2), 'utf-8');

    // Atomic rename
    renameSync(tempFile, configFile);
  } finally {
    await release();
  }
}

/**
 * Read, update, and write config atomically.
 */
export async function updateConfig(
  updater: (config: Config) => Config
): Promise<Config> {
  const configFile = getConfigFile(true);

  // Ensure file exists before locking
  if (!existsSync(configFile)) {
    writeFileSync(configFile, JSON.stringify(createDefaultConfig(), null, 2));
  }

  const release = await lockfile.lock(configFile, {
    retries: { retries: 3, minTimeout: 100, maxTimeout: 1000 },
    stale: LOCK_TIMEOUT,
    realpath: false,
  });

  try {
    const config = readConfig();
    const updated = updater(config);

    // Write to temp file first
    const tempFile = configFile.replace('.json', '.tmp');
    writeFileSync(tempFile, JSON.stringify(updated, null, 2), 'utf-8');

    // Atomic rename
    renameSync(tempFile, configFile);

    return updated;
  } finally {
    await release();
  }
}

File: services/config/src/debrief_config/storage.py
"""Config file storage with atomic writes and file locking.

Provides safe concurrent access to the config file using:
1. File locking to prevent simultaneous writes
2. Atomic writes via temp file + rename to prevent corruption
"""

import json
import logging

from filelock import FileLock, Timeout

from .models import Config
from .paths import get_config_file, get_lock_file

logger = logging.getLogger(__name__)

# Lock timeout in seconds
LOCK_TIMEOUT = 5.0


def read_config() -> Config:
    """Read the configuration file.

    If the file doesn't exist, returns default config.
    If the file is corrupted, logs a warning and returns default config.

    Returns:
        Config object with current settings.
    """
    config_file = get_config_file(ensure_dir_exists=True)

    if not config_file.exists():
        logger.debug("Config file not found, using defaults")
        return Config.default()

    try:
        data = json.loads(config_file.read_text(encoding="utf-8"))
        return Config.model_validate(data)
    except json.JSONDecodeError as e:
        logger.warning(f"Config file corrupted (invalid JSON): {e}")
        return Config.default()
    except Exception as e:
        logger.warning(f"Config file validation failed: {e}")
        return Config.default()


def write_config(config: Config) -> None:
    """Write the configuration file atomically with locking.

    Uses a lock file to prevent concurrent writes, and writes to a
    temp file first then renames for atomic operation.

    Args:
        config: Config object to write.

    Raises:
        Timeout: If lock cannot be acquired within timeout.
    """
    config_file = get_config_file(ensure_dir_exists=True)
    lock_file = get_lock_file()

    lock = FileLock(lock_file, timeout=LOCK_TIMEOUT)

    try:
        with lock:
            # Write to temp file first
            temp_file = config_file.with_suffix(".tmp")

            # Serialize with aliases for camelCase field names
            json_data = config.model_dump(mode="json", by_alias=True)
            temp_file.write_text(
                json.dumps(json_data, indent=2, ensure_ascii=False),
                encoding="utf-8",
            )

            # Atomic rename (POSIX) or replace (Windows)
            temp_file.replace(config_file)

            logger.debug(f"Config written to {config_file}")

    except Timeout:
        logger.error(f"Could not acquire lock on {lock_file}")
        raise


def update_config(updater: callable) -> Config:
    """Read, update, and write config atomically.

    Args:
        updater: Function that takes Config and returns modified Config.

    Returns:
        Updated Config object.
    """
    config_file = get_config_file(ensure_dir_exists=True)
    lock_file = get_lock_file()
    lock = FileLock(lock_file, timeout=LOCK_TIMEOUT)

    try:
        with lock:
            # Read current config
            config = read_config()
            updated = updater(config)

            # Write directly (don't call write_config to avoid nested lock)
            temp_file = config_file.with_suffix(".tmp")
            json_data = updated.model_dump(mode="json", by_alias=True)
            temp_file.write_text(
                json.dumps(json_data, indent=2, ensure_ascii=False),
                encoding="utf-8",
            )
            temp_file.replace(config_file)
            logger.debug(f"Config updated in {config_file}")

            return updated
    except Timeout:
        logger.error(f"Could not acquire lock on {lock_file}")
        raise
